### PR TITLE
nixos/anki-sync-server: init

### PR DESCRIPTION
**Pull request by telotortium**
## Description of changes
Provide a NixOS module for the [built-in Anki Sync Server](https://docs.ankiweb.net/sync-server.html) included in recent versions of Anki. This supersedes the `ankisyncd` module, but we should keep that for now because `ankisyncd` supports older versions of Anki clients than this module.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by telotortium** [comment]
@martinetd as maintainer of ankisyncd.

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (de11n) [inline nixos/modules/services/misc/anki-sync-server.nix:136]: Instead of embedding this directly in the unit, would it make more sense to use `writeScriptBin` for this and just use that with `ExecStart`? Makes it easier to run the same code outside of systemd.

**Review by de11n — COMMENTED** [review]
I don't know anything about Anki so I just have random stylistic comments.

**Comment by telotortium** [comment]
@de11n in the manual test, why isn't my introduced option `services.anki-sync-server.enable` producing a correct reference `#opt-services.anki-sync-server.enable`?

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (de11n) [inline nixos/modules/services/misc/anki-sync-server.nix:135]: There's a lot of "dynamic" programming in here. E.g. in Nix you know what variables need to be populated from files and which don't. Instead of rendering env vars and then writing complex bash to parse it, why not just render the bash directly to do the correct thing? That would get rid of the gotcha about passwords starting with `/`.

Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix]: Done

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix]: Done

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (martinetd) [inline nixos/modules/services/misc/anki-sync-server.nix:40]: Would be great if someone would split the built-in server:
https://docs.ankiweb.net/sync-server.html#with-cargo

That should generate a much smaller package (without qt etc dependencies)

Doesn't have to be done in this PR, but that was one of the reasons I didn't do this back when I added anki-sync-server-rs (the standalone server didn't exist at the time)

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (martinetd) [inline nixos/modules/services/misc/anki-sync-server.nix:119]: (does this make sense in nixpkgs? To avoid this commit being backported? my impression was that new modules aren't usually backported to stable trees)

File: nixos/modules/services/misc/anki-sync-server.md
Comment (martinetd) [inline nixos/modules/services/misc/anki-sync-server.md:68]: I'm curious where did you see that? It works fine with 2.1.66 and I already have it patched for the 23.10 beta that does require an update; so it seems fine to use as far as I'm concerned:
https://github.com/ankicommunity/anki-sync-server-rs/pull/76
(I haven't submitted it to nixpkgs yet because it requires https://github.com/ankitects/rust-csv/pull/1 or equivalent to build anki itself and I didn't get any feedback from upstream yet; but I have it all ready along with a trivial nixosTest...)

We were discussing the future of anki-sync-server-rs vs the built-in server in an old PR here: https://github.com/NixOS/nixpkgs/pull/224366#issuecomment-1714697175

I guess since you've done the work now we can keep both in parallel for a while; I was mostly curious about the compatibility of the server data since users aren't too bad to configure again.

**Review by martinetd — COMMENTED** [review]
Thanks for your work! Everything about packaging anki has been a pain to me, so anyone working on this is greatly appreciated.
I've got a few high level questions first, didn't look at the module itself yet; I'll take a bit of time for that this weekend.

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (de11n) [inline nixos/modules/services/misc/anki-sync-server.nix:29]:
```suggestion
      (x: ''export SYNC_USER${toString x.i}=${escapeShellArg x.user.username}:"''$(cat "''${CREDENTIALS_DIRECTORY}/"${escapeShellArg x.user.username})"'')
```

**Review by de11n — COMMENTED** [review]
Much better! Thanks! I'll let others take it from here who actually know what Anki does. 😄

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix:40]: Noted

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix:119]: @de11n Do you think I should take this out because the version in master nixpkgs works? I put this in originally because I found it useful on my machine.

File: nixos/modules/services/misc/anki-sync-server.md
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.md:68]: Ah, I got that impression based on the anki-sync-server-rs README: https://github.com/ankicommunity/anki-sync-server-rs/blob/5280f9f2034b29d77388df497ebcdb7d13cb5c94/README.md

Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.md:68]: For now, we should probably keep them both, until we figure out how to migrate data between the two servers in a less painful way.

Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.md:68]: Right, that note in the README is what got us started in the nixpkgs thread -- I didn't realize it also explicitly mentioned 1.64.

As far as the protocol (and thus clients supported) go, anki-sync-server-rs uses the code from anki itself so they support exactly the same thing.
As long as someone keeps rebasing the small anki patch in anki-sync-server-rs I don't expect any breakage in the short-medium term; I've spent much more time getting anki 23.10beta1 to build on nixpkgs (because of the cargo vendor thing) than updating the standalone server patch :|

Long-term though I agree the standalone server is fragile so migrating to this makes more sense; when writing the nixosTest for ankisyncd I also painfully realized how non-nixos-y the old server is (having to run a command manually in the data dir to manage users), but I'd really like to be able to preserve data itself.

I'll try to run the built in anki server and compare data directories this weekend

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (de11n) [reply nixos/modules/services/misc/anki-sync-server.nix:119]: I don't know the answer to this. I think it might depend on how common/likely we think it will be for users to configure their own version of `anki`. If there's some reason to believe that will be common, then it makes sense to help users get it right. If we expect folks to rarely, if ever, configure this manually, then it's probably overkill. NixOS tests would be ideal to catch breakages.

File: nixos/modules/services/misc/anki-sync-server.md
Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.md:68]: So the build-in server data seems compatible: just need to set the base path (SYNC_BASE) to /var/lib/ankisyncd/collections and it rolls.
The login token isn't compatible -- I get an auth error on switching from ankisyncd to anki's server and back, but data itself seems fine, which is what's most important.

With that I guess we really ought to make the switch, there's really no point in keeping both... I have mixed feelings as I just did the port to anki-sync-server-rs but rationally it's definitely for the best. There have been no release with the new server, so I'd go ahead and replace it, just keeping the SYNC_BASE for convenience.
(sorry for the delay, and rushing this in after it...)

I'll see about packaging the standalone server either this week or next week; anki 23.10 should now build fine with https://github.com/ankitects/anki/pull/2703 hopefully.
I'll also test this PR a bit more thoroughly, I've never used systemd's LoadCredential so I just don't get it without testing, but if it works it looks good to me.

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.nix:40]: Done in https://github.com/NixOS/nixpkgs/pull/261166 (Cc'd you directly)

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.nix:119]: Hmm, this flat out didn't work for me:
```
error:
       … while evaluating the attribute 'drvPath'
         at /home/user/nixpkgs/lib/customisation.nix:222:7:
          221|     in commonAttrs // {
          222|       drvPath = assert condition; drv.drvPath;
             |       ^
          223|       outPath = assert condition; drv.outPath;
       … while calling the 'derivationStrict' builtin
         at /builtin/derivation.nix:9:12: (source not available)
       (stack trace truncated; use '--show-trace' to show the full trace)
       error: value is null while a list was expected
```
Very helpful error and --show-trace didn't help as usual; sometimes nix is just bad at this... I've flat out removed the check in my test branch:
https://github.com/martinetd/nixpkgs/commit/bd6a4753f10ce96ea61db4417cddc8a1b7212b0b
but feel free to amend your commits

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (martinetd) [inline nixos/modules/services/misc/anki-sync-server.nix:55]: This doesn't work:
```
Sync server failed: reading SYNC_* env vars

Caused by this error:
  1: invalid IP address syntax
```
I've changed the default to `"::"` in https://github.com/martinetd/nixpkgs/commit/52ebc9fef2f55363b3e6521f83009527c97354d1 but the comment also needs updating; feel free to not take this commit

File: nixos/modules/services/misc/anki-sync-server.md
Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.md:68]: Ok on LoadCredential; this is more complicated than I'd do (I'd just ask users to fill in an EnvironmentFile themselves), but it works and is certainly more nixos-y so that's fine with me

**Review by martinetd — CHANGES_REQUESTED** [review]
I've played with this a bit, rebased and added a few commits in https://github.com/martinetd/nixpkgs/commits/anki-sync-server-nixos

I didn't have anywhere to cmment individually, but this is missing adding the module to `nixos/modules/module-list.nix` ( https://github.com/martinetd/nixpkgs/commit/ccfbafec405b4671a384db348fdde3e69b91fa9d ); which makes this hard to use :) Took me a while to understand why test didn't find `services.anki-sync-server` !

I've added a nixosTests and switched to the pending anki-sync-server binary; not quite sure what to do order-wise here wrt this PR, but it's identical to anki --syncserver so this is probably fine to do later.
(We might want to change `services.anki-sync-server.package` in favor of something like `server_command` that'd be either `"${pkgs.anki}/bin/anki --syncserver"` or `"${pkgs.anki-sync-server}/bin/anki-sync-server"` ...)

~~The added test *needs* the 23.10beta6 update because anki's internal changed a tiny little bit (at least `col.sync_collection(login, False)` needs to be changed to `col.sync_collection(login)` for 2.1.66); I think rest would work)~~
The tests need updating for 23.10 so I've split the test commit and its update as we probably don't want to wait for anki update (the anki-sync-server package looks like it'll get in fast enough to use)


Last would be to really decide how we want the modules to evolve.
Keeping both services.ankisyncd and services.anki-sync-server might seem good at first but data migration will be a pain later because we'll have users with data in either directory: I'd be in favor of replacing services.ankisyncd with this, keeping just the data path working as said last week.

WDYT?

**Comment by telotortium** [comment]
@martinetd sorry for not getting back earlier - I didn't realize you had committed so many changes on your branch. I've pushed your commits to this PR.

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix]: We removed this warning, because upstream is now using Anki 23.10.

**Comment by telotortium** [comment]
@martinetd do you still want to merge #261166? I've integrated those changes and added a few fixes here as well, so I think this branch should be merged.

**Comment by telotortium** [comment]
> @de11n in the manual test, why isn't my introduced option `services.anki-sync-server.enable` producing a correct reference `#opt-services.anki-sync-server.enable`?

Figured it out - I had to add my new module to nixos/modules/module-list.nix

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix:40]: Let's decide whether to merge this one or the other.

Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.nix:40]: I think it's easier to deal with small PRs (that one, a PR to update anki to 23.10, then this) but either work for me; I wanted to try to get this in small bits before nixos 23.11 but I think we're a wee late now so either work for me :)

For this PR as of now, we can't take https://github.com/NixOS/nixpkgs/pull/257692/commits/d7ae7c637c487147f83019752864e0a29fa8bbe7 (update anki to 23.10beta6)  -- I didn't keep updating the old branch but we'll need it to be a stable version (request from anki maintainers in the other PR, but it totally makes sense); anki 23.10.1 got released since so we can go ahead and upgrade to that.
Do you want to give it a try or shall I?

As for which PR to merge, I really don't mind either way; if this is ready I'm happy to close the other one. Otherwise I'll post it to the discourse topic of PRs ready to be merged and it shouldn't take too long...

Comment (martinetd) [reply nixos/modules/services/misc/anki-sync-server.nix:40]: Haven't done anything for anki 23.10.1, but I just did 23.10beta6 -> 23.10.1 for anki-sync-server-rs's anki dependency as a last struggle without any trouble so I expect it'll be fine just updating hashes (src and cargo deps).
I'll let you update this and close the other PR

**Review by martinetd — COMMENTED** [review]
(Last review before you update the package -- as far as I'm concerned, the new module is good to go, just these small details + anki update left)

**Comment by telotortium** [comment]
@martinetd Can we drop the initial commit with 2.1.66 and directly start at 23.10 (ideally 23.10.1)? I presume so

**Comment by martinetd** [comment]
> @martinetd Can we drop the initial commit with 2.1.66 and directly start at 23.10 (ideally 23.10.1)? I presume so

Sure feel free to rewrite history, but the commit should be separate as it updates anki too, not just anki-sync-server.
So I guess you can reoder the commits and squash the test fix, but that's about it?
I have no hard preference, feel free to do as you like :)

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix]: I updated to Anki 23.10.1 in this PR.

**Comment by telotortium** [comment]
> > @martinetd Can we drop the initial commit with 2.1.66 and directly start at 23.10 (ideally 23.10.1)? I presume so
> 
> Sure feel free to rewrite history, but the commit should be separate as it updates anki too, not just anki-sync-server. So I guess you can reoder the commits and squash the test fix, but that's about it? I have no hard preference, feel free to do as you like :)

I started with 2.1.66 and make a separate commit to update to 23.10.1.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (telotortium) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: Maybe we can update ankisyncd in a separate smaller PR - I want to get this one in.

File: nixos/modules/services/misc/anki-sync-server.nix
Comment (telotortium) [reply nixos/modules/services/misc/anki-sync-server.nix]: I updated to 23.10.1

**Comment by telotortium** [comment]
@de11n @martinetd can one of you merge?

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (telotortium) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: Done

**Comment by paveloom** [comment]
I would suggest making this dependent on https://github.com/NixOS/nixpkgs/pull/264796 and marking as a draft until that PR is merged.

**Review by martinetd — COMMENTED** [review]
> I would suggest making this dependent on https://github.com/NixOS/nixpkgs/pull/264796 and marking as a draft until that PR is merged.

It's been open for a while but I guess it's likely the other PR will get in first...
We don't actually really care about which version gets in except for the test that breaks; we could just remove the version update and test addition for now and add it in a later PR @telotortium ?
That way this PR won't need to wait on anything.

And I do not have merge rights, sorry. Once the release note comment is fixed up we can bring it up in https://discourse.nixos.org/t/prs-already-reviewed/2617 and someone should take a look.

**Comment by martinetd** [comment]
@wegank merged both the anki update and the anki-sync-server package addition just now (Thanks!!)

I'll rebase this and fix the release notes wording tonight if you don't have time before me; shouldn't be much longer :)

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (wegank) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: This should probably be done in a PR deprecating the module, I think.

**Review by martinetd — APPROVED** [review]

**Comment by telotortium** [comment]
@wegank can we merge?

**Pull request merged**
